File: src/chipschallenge/blockreactions/WaterFrom.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;

public class WaterFrom extends BlockReaction {

    private WaterFrom() {
    }
    private static WaterFrom mInstance = null;

    public static synchronized WaterFrom getInstance() {
        if (mInstance == null) {
            mInstance = new WaterFrom();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            moving.setType(Type.CHIP);
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/WaterFrom.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Type = Block.Type;

    public class WaterFrom : NoSlipReaction
    {
        private static WaterFrom instance;

        private WaterFrom()
        {
        }

        public static WaterFrom Instance
        {
            get
            {
                lock (typeof(WaterFrom))
                {
                    return instance ?? (instance = new WaterFrom());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                moving.SetType(Type.CHIP);
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}